constantine/math_compiler/pub_fields.nim

```diff
@@ -6,12 +6,198 @@
+  let name = fd.name & "_setZero_internal"
+  asy.llvmInternalFnDef(
+          name, SectionName,
+          asy.void_t, toTypes([r]),
+          {kHot}):
+    tagParameter(1, "sret")
+    let M = asy.getModulusPtr(fd)
+
+    let ri = llvmParams
+    let rA = asy.asField(fd, ri)
+    for i in 0 ..< fd.numWords:
+      rA[i] = constInt(fd.wordTy, 0)
+
+    asy.br.retVoid()
+
+  asy.callFn(name, [r])
+
+proc genFpSetZero*(asy: Assembler_LLVM, fd: FieldDescriptor): string =
+  ## Generate a public field setZero procedure
+  ## with signature
+  ##   void name(FieldType r)
+  ## with r the element to be zeroed.
+  ## and return the corresponding name to call it
+
+  let name = fd.name & "_setZero"
+  asy.llvmPublicFnDef(name, "ctt." & fd.name, asy.void_t, [fd.fieldTy]):
+    let r = llvmParams
+    asy.setZero_internal(fd, r)
+    asy.br.retVoid()
+
+  return name
+
+proc setOne_internal*(asy: Assembler_LLVM, fd: FieldDescriptor, r: ValueRef) {.used.} =
+  ## Generate an internal field setOne
+  ## with signature
+  ##   void name(FieldType r)
+  ## with `r` the element to be set to 1 in Montgomery form.
+  ##
+  ## Generates a call, so that we one can use this proc as part of another procedure.
+  let name = fd.name & "_setOne_internal"
+  asy.llvmInternalFnDef(
+          name, SectionName,
+          asy.void_t, toTypes([r]),
+          {kHot}):
+    tagParameter(1, "sret")
+    let M = asy.getModulusPtr(fd)
+    let ri = llvmParams
+    let rF = asy.asField(fd, ri)
+
+    let mOne = asy.getMontyOnePtr(fd)
+    let mF = asy.asField(fd, mOne)
+
+    # Need to call `store` for `Field`, which actually copies
+    # the full data!
+    store(rF, mF)
+
+    asy.br.retVoid()
+
+  asy.callFn(name, [r])
+
+proc genFpSetOne*(asy: Assembler_LLVM, fd: FieldDescriptor): string =
+  ## Generate a public field setOne procedure
+  ## with signature
+  ##   void name(FieldType r)
+  ## with `r` the element to be set to 1 in Montgomery form.
+  ## and return the corresponding name to call it
+
+  let name = fd.name & "_setOne"
+  asy.llvmPublicFnDef(name, "ctt." & fd.name, asy.void_t, [fd.fieldTy]):
+    let r = llvmParams
+    asy.setOne_internal(fd, r)
+    asy.br.retVoid()
+
+  return name
+
+proc add_internal*(asy: Assembler_LLVM, fd: FieldDescriptor, r, a, b: ValueRef) {.used.} =
+  ## Generate an internal field addition proc
+  ## with signature
+  ##   void name(FieldType r, FieldType a, FieldType b)
+  ## with r the result and a, b the operands
+  ##
+  ## Generates a call, so that we one can use this proc as part of another procedure.
+  let name = fd.name & "_add_internal"
+  asy.llvmInternalFnDef(
+          name, SectionName,
+          asy.void_t, toTypes([r, a, b]),
+          {kHot}):
+    tagParameter(1, "sret")
+    let M = asy.getModulusPtr(fd)
+
+    let (ri, ai, bi) = llvmParams
+    asy.modadd(fd, ri, ai, bi, M)
+    asy.br.retVoid()
+
+  asy.callFn(name, [r, a, b])
```

Comment:
Ah, I see, there are 3 types of functions:
- underlying primitives
- internal
- public

and for fields, `internal` is just forwarding to modular arithmetic primitives

In the past I created an enum with the types of functions: https://github.com/mratsim/constantine/blob/cfe077b682b625b00d2e95693e8753570981d278/constantine/math_compiler/ir.nim#L202-L210

and I was planning to associate (opcode, curve) -> primitive, to call the proper underlying implementation.

I have yet to find the best solution but I think remove the _internal noise would be nice to improve debugging, perf optimization and dev experience.
Having all those `_internal` is annoying and unlike `_vartime` there is no security risk to justify it.

I think they should all be moved to the impl_*.nim file and renamed without `_internal`.

Public functions wont collide in LLVM due to the `ctt_` prefix and the Nim ones won't collide because the internal functions should not be reexported. Which is the case if they are implemented in the `pub_*.nim` like now.